### DISCUSSION TITLE
camera acquisition fps issue (with DetectPose from DLC)

### DISCUSSION THREAD

**Discussion by thedennislab** [bonsai-rx/bonsai#1078] [Q&A]
hi folks,

We have a basler acA2040-90umNIR camera. We want to use the data from this for closed-loop behavior and tracking. We can do this really successfully, but our fps in saving out is quite low. To debug we made a very simplified workflow, see below. without the DetectPose node, we get 90fps (~0.012 sec per frame). With it, we get variable results from 7-30"fps" (so anything from .2-.03 sec per frame). Bruno helpfully suggested we try adding a publish subject after video writer. This also did not solve, see below for the workflow we tried.

I don't understand why the /acquisition/ is affected here. We need to save out our videos at full fps. We also want to use that video for processing and tracking. I think this is something I'm missing about how Rx works but it's beyond me for now. Thanks for any/all help! -ejd
![dlcissue_2](https://user-images.githubusercontent.com/67440835/195383687-b4d6900f-a59f-4a14-9a45-fe61fcd90fe9.svg)
![dlcissue](https://user-images.githubusercontent.com/67440835/195383690-e8959350-d4b0-460f-8fd0-0b516943226b.svg)

**Comment by bruno-f-cruz** [comment]
Can you attach your workflows? Thanks!

**Reply by thedennislab** [reply]
[dlc_fps_issue.zip](https://github.com/bonsai-rx/bonsai/files/9767795/dlc_fps_issue.zip)

**Reply by thedennislab** [reply]
[dlc_fps_issue_basic.zip](https://github.com/bonsai-rx/bonsai/files/9767800/dlc_fps_issue_basic.zip)

**Reply by thedennislab** [reply]
for both, there are variable and long times between each frame, up to .13 seconds (7fps). For both, removing or disabling the DetectPose node results in low latencies and the expected frame rate (90fps). It makes sense to me that things downstream of DetectPose are slow (and that is fine for our needs) but we want to separately acquire these images at full 90 fps and save them out.

**Reply by bruno-f-cruz** [reply]
It just looks like your machine is not able to process the frames fast enough and you are likely dropping frames. (see a longer explanation here https://groups.google.com/g/bonsai-users/c/vyyKsKJ9UR0/m/neatze1zBQAJ)

I thought a publishsubject would do the trick by decoupling the streams but i guess not :P maybe Gonçalo has a better option but, in the meanwhile, you can decrease the number of frames arriving to the inference node by using the `Slice` operator effectively downsampling the fps after the videowriter node.

**Reply by thedennislab** [reply]
with the 2.7 release, we tried (as @bruno-f-cruz and @glopesdev suggested) to use a skip between camera and DLC, and a ObserveOn (TaskPoolScheduler) node just before Predict Pose. This did fix the camera saving out frame rate issue (now easily and consistently at 90fps) but DLC takes the first few frames of the video and loops over them, finding the post estimate of the first few frames over and over again. Since we're trying to use DLC-live for live inference to drive a Zaber stage to stay over the animal at all times, this is of course not ideal. 

Attaching the zipped workflow we're using to debug. Any/all help appreciated, I think something like this will be a very common use case and solving it hopefully helps the community too. 
[debug_DLC_cameraCapture_issue_20221031.zip](https://github.com/bonsai-rx/bonsai/files/9902007/debug_DLC_cameraCapture_issue_20221031.zip)

**Comment by jonnew** [comment]
I just want to get some simple insight here. Do we have a feeling as to the root cause of this? e.g. is this because the detect node pose is simply a huge bottleneck and cannot process frames as fast as they arrive? If so, is it a hardware issue (will not be solved without upgraded GPU), software issue (underlying DLC implementation), or DLC "API" usage (if some buffering were to occur before interacting with DLC API things would improve).

**Reply by thedennislab** [reply]
We've never gone above 12% GPU nor CPU usage, so I don't think this is a hardware issue. This computer has 128Gb RAM and two NVIDIA GeForce 2080 Tis

**Reply by thedennislab** [reply]
we also don't care if we need to downsample in time for detect pose, our tracking works well down to 20fps (tried using Skips for this and it also didn't fix anything, DLC still loops frames). We just want to also save out our videos at 90fps.

We also decreased the acquisition fps to 2 and still detect pose loops over the first few frames.

For now we're just accepting the loss of frames and irregularity in our data (camera frames get saved out every .1-.7 seconds or so with a timestamp) so we can get some things done but this won't work for us long term and would mean we would stop using Bonsai for our data acquisition... which we don't want to do!
